### DISCUSSION TITLE
ejabberd document NOT updated : IQ Handler discipline

### DISCUSSION THREAD

**Discussion by AobaIwaki123** [processone/ejabberd#4129] [Questions] [answered]
# About
I read about [IQ descipline](https://docs.ejabberd.im/developer/guide/#iq-discipline) to understand IQ handler  
It says IQ handler supports parallel execution 
However it's deprecated from [gen_iq_handler.erl](https://github.com/processone/ejabberd/blob/master/src/gen_iq_handler.erl) in master branch 
# Question
I want to know how to make IQ handler parallel

**Comment by badlop** [comment, marked as answer]
Right, that was deprecated in ejabberd 18.03. The [commit message](https://github.com/processone/ejabberd/commit/66fc1bf3b6cab63559fa58a086a1439ecdebfea7) explains it.

And right, that paragraph in the documentation is obsolete ans should be removed.

Thanks for reporting it!

**Reply by AobaIwaki123** [reply]
Thank you for reply!
Anyway, is it possible to execute iq handler in parallel?

**Reply by badlop** [reply]
**Update:** No need to do anything, iq queries are handled in parallel already.

I applied the patch (1) to confirm that IQ queries are processed in parallel.

This patch adds a 10 seconds delay to IQ disco-features queries sent by user1; other users have no delay.

When user1 sends a query, he has to wait 10 seconds to receive the response from ejabberd; in the meantime, any other user can send similar queries and they are processed immediately, without waiting for user1:

    # user1 sends query but doesn't yet receive the response:
    2024-09-02 10:28:30.698067+02:00 [info] mod_disco proces_local_iq_info start <<"user1">>
    2024-09-02 10:28:30.698225+02:00 [info] mod_announce disco features start <<"user1">>
    
    # admin sends query and receives the response immediately:
    2024-09-02 10:28:33.654562+02:00 [info] mod_disco proces_local_iq_info start <<"admin">>
    2024-09-02 10:28:33.654704+02:00 [info] mod_announce disco features start <<"admin">>
    2024-09-02 10:28:33.654854+02:00 [info] mod_announce disco features result <<"admin">>
    2024-09-02 10:28:33.655111+02:00 [info] mod_disco proces_local_iq_info result <<"admin">>
    
    # user1 finall receives the response:
    2024-09-02 10:28:40.698818+02:00 [info] mod_announce disco features result <<"user1">>
    2024-09-02 10:28:40.699266+02:00 [info] mod_disco proces_local_iq_info result <<"user1">>


(1) Patch for ejabberd git:

    diff --git a/src/mod_announce.erl b/src/mod_announce.erl
    index d7f983716..9aca6cd8d 100644
    --- a/src/mod_announce.erl
    +++ b/src/mod_announce.erl
    @@ -262,6 +262,7 @@ disco_features(Acc, From, #jid{lserver = LServer} = _To, <<"announce">>, Lang) -
         end;
     
     disco_features(Acc, From, #jid{lserver = LServer} = _To, Node, Lang) ->
    +    ?INFO_MSG("mod_announce disco features start ~p", [From#jid.luser]),
         case gen_mod:is_loaded(LServer, mod_adhoc) of
     	false ->
     	    Acc;
    @@ -270,6 +271,11 @@ disco_features(Acc, From, #jid{lserver = LServer} = _To, Node, Lang) ->
     	    Allow = acl:match_rule(LServer, Access, From),
     	    AccessGlobal = get_access(global),
     	    AllowGlobal = acl:match_rule(global, AccessGlobal, From),
    +            case From#jid.luser of
    +                <<"user1">> -> timer:sleep(10000);
    +                _ -> ok
    +            end,
    +            ?INFO_MSG("mod_announce disco features result ~p", [From#jid.luser]),
     	    case Node of
     		?NS_ADMIN_ANNOUNCE ->
     		    ?INFO_RESULT(Allow, [?NS_COMMANDS], Lang);
    diff --git a/src/mod_disco.erl b/src/mod_disco.erl
    index 4f092da4e..f1248b3f2 100644
    --- a/src/mod_disco.erl
    +++ b/src/mod_disco.erl
    @@ -119,6 +119,7 @@ process_local_iq_info(#iq{type = set, lang = Lang} = IQ) ->
     process_local_iq_info(#iq{type = get, lang = Lang,
     			  from = From, to = To,
     			  sub_els = [#disco_info{node = Node}]} = IQ) ->
    +    ?INFO_MSG("mod_disco proces_local_iq_info start ~p", [From#jid.luser]),
         Host = To#jid.lserver,
         Identity = ejabberd_hooks:run_fold(disco_local_identity,
     				       Host, [], [From, To, Node, Lang]),
    @@ -127,6 +128,7 @@ process_local_iq_info(#iq{type = get, lang = Lang,
         case ejabberd_hooks:run_fold(disco_local_features, Host,
     				 empty, [From, To, Node, Lang]) of
     	{result, Features} ->
    +            ?INFO_MSG("mod_disco proces_local_iq_info result ~p", [From#jid.luser]),
     	    xmpp:make_iq_result(IQ, #disco_info{node = Node,
     						identities = Identity,
     						xdata = Info,


  [1]: https://github.com/processone/ejabberd/commit/66fc1bf3b6cab63559fa58a086a1439ecdebfea7
